### PR TITLE
Fixed timers in upload

### LINKED ISSUE #31: Plugin neve catch up with upload demand
I'm using the plugin to get logs from HAProxy into the BigQuery tables. I can parse the lines using Grok and send the parsed data to the plugin.

My problem is that the plugin can't seem to handle uploads efficiently. The temp files keeps growing in number and eventually fills up the disk space. Here is my config:

```
flush_interval_secs => 2
uploader_interval_secs => 300
deleter_interval_secs => 2
```

It seems the uploader_interval_secs is used to stop appending data to one part file, and start a new one (as you can see below one file every 5 mins), while uploading the previous one. As the amount of files keep growing, I think the trigger for uploading is not working correctly or should be a separate configuration.

Here is a list of files that are not being uploaded as fast as it should:
```
total 1.2G
-rw-r--r-- 1 root root  42M Sep 11 12:35 logstash_bq_61e4914c872e_2017-09-11T12:00.part006.log
-rw-r--r-- 1 root root  44M Sep 11 12:40 logstash_bq_61e4914c872e_2017-09-11T12:00.part007.log
-rw-r--r-- 1 root root  38M Sep 11 12:45 logstash_bq_61e4914c872e_2017-09-11T12:00.part008.log
-rw-r--r-- 1 root root  37M Sep 11 12:50 logstash_bq_61e4914c872e_2017-09-11T12:00.part009.log
-rw-r--r-- 1 root root  39M Sep 11 12:55 logstash_bq_61e4914c872e_2017-09-11T12:00.part010.log
-rw-r--r-- 1 root root  36M Sep 11 13:00 logstash_bq_61e4914c872e_2017-09-11T12:00.part011.log
-rw-r--r-- 1 root root  39M Sep 11 13:05 logstash_bq_61e4914c872e_2017-09-11T13:00.part000.log
-rw-r--r-- 1 root root  37M Sep 11 13:10 logstash_bq_61e4914c872e_2017-09-11T13:00.part001.log
-rw-r--r-- 1 root root  36M Sep 11 13:15 logstash_bq_61e4914c872e_2017-09-11T13:00.part002.log
-rw-r--r-- 1 root root  38M Sep 11 13:20 logstash_bq_61e4914c872e_2017-09-11T13:00.part003.log
-rw-r--r-- 1 root root  39M Sep 11 13:25 logstash_bq_61e4914c872e_2017-09-11T13:00.part004.log
-rw-r--r-- 1 root root  35M Sep 11 13:30 logstash_bq_61e4914c872e_2017-09-11T13:00.part005.log
-rw-r--r-- 1 root root  34M Sep 11 13:35 logstash_bq_61e4914c872e_2017-09-11T13:00.part006.log
-rw-r--r-- 1 root root  37M Sep 11 13:40 logstash_bq_61e4914c872e_2017-09-11T13:00.part007.log
-rw-r--r-- 1 root root  38M Sep 11 13:45 logstash_bq_61e4914c872e_2017-09-11T13:00.part008.log
-rw-r--r-- 1 root root  37M Sep 11 13:50 logstash_bq_61e4914c872e_2017-09-11T13:00.part009.log
-rw-r--r-- 1 root root  39M Sep 11 13:55 logstash_bq_61e4914c872e_2017-09-11T13:00.part010.log
-rw-r--r-- 1 root root  38M Sep 11 14:00 logstash_bq_61e4914c872e_2017-09-11T13:00.part011.log
-rw-r--r-- 1 root root  40M Sep 11 14:05 logstash_bq_61e4914c872e_2017-09-11T14:00.part000.log
-rw-r--r-- 1 root root  40M Sep 11 14:10 logstash_bq_61e4914c872e_2017-09-11T14:00.part001.log
-rw-r--r-- 1 root root  39M Sep 11 14:15 logstash_bq_61e4914c872e_2017-09-11T14:00.part002.log
-rw-r--r-- 1 root root  37M Sep 11 14:20 logstash_bq_61e4914c872e_2017-09-11T14:00.part003.log
-rw-r--r-- 1 root root  40M Sep 11 14:25 logstash_bq_61e4914c872e_2017-09-11T14:00.part004.log
-rw-r--r-- 1 root root  39M Sep 11 14:30 logstash_bq_61e4914c872e_2017-09-11T14:00.part005.log
-rw-r--r-- 1 root root  39M Sep 11 14:35 logstash_bq_61e4914c872e_2017-09-11T14:00.part006.log
-rw-r--r-- 1 root root  40M Sep 11 14:40 logstash_bq_61e4914c872e_2017-09-11T14:00.part007.log
-rw-r--r-- 1 root root  39M Sep 11 14:45 logstash_bq_61e4914c872e_2017-09-11T14:00.part008.log
-rw-r--r-- 1 root root  37M Sep 11 14:50 logstash_bq_61e4914c872e_2017-09-11T14:00.part009.log
-rw-r--r-- 1 root root  38M Sep 11 14:55 logstash_bq_61e4914c872e_2017-09-11T14:00.part010.log
-rw-r--r-- 1 root root  37M Sep 11 15:00 logstash_bq_61e4914c872e_2017-09-11T14:00.part011.log
-rw-r--r-- 1 root root  36M Sep 11 15:05 logstash_bq_61e4914c872e_2017-09-11T15:00.part000.log
-rw-r--r-- 1 root root  39M Sep 11 15:10 logstash_bq_61e4914c872e_2017-09-11T15:00.part001.log
-rw-r--r-- 1 root root 2.4M Sep 11 15:10 logstash_bq_61e4914c872e_2017-09-11T15:00.part002.log
```

I can try getting a PR up, but need some guidance on the approach to fix this.

### PR DESCRIPTION
**Pull request by josephlewis42**
There was a bug that the uploaders may never catch up because they
were pausing even if multiple files were waiting in the
queue to be uploaded. This problem could be compounded if the
plugin attempted a recovery and started with many items in the
queue.

The solution is to only delay if we have no pending uploads,
otherwise we keep uploading. It should fix #31

### CODE REVIEW

File: lib/logstash/outputs/google_bigquery.rb [hunk 1]
```diff
@@ -411,9 +424,13 @@ def initialize_uploader
...
               @upload_queue << filename
+
               # If we got here, it means that older files were uploaded, so let's
-              # wait another minute before checking on this file again.
-              sleep @uploader_interval_secs
+              # wait before checking on this file again.
+              #
+              # Use the min so we don't wait too long if the logs start rotating too quickly
+              # due to an increased number of events.
+              sleep [60, @uploader_interval_secs].min
```
Comment (colinsurprenant) [inline lib/logstash/outputs/google_bigquery.rb]: not sure why you cap it at 60 here?

**Comment by colinsurprenant** [comment]
Thanks for you contribution @josephlewis42.

The logic is quite hard to follow in the upload section. IIUC, it was sleeping after each `@upload_queue.pop`? Was there a reason for wanting to do so?

You proposal is to kill that systematic sleep and only keep the one where we need to re-enqueue - I don't understand that logic ...

File: lib/logstash/outputs/google_bigquery.rb (hunk 1 above)
Comment (josephlewis42) [reply lib/logstash/outputs/google_bigquery.rb]: Based on a lot of the backlog issues around race conditions I didn't fully trust that there weren't other files in the queue if the head was the current file so I capped it down to the interval that the original comment said.

It's weird that the upload queue is used both for message passing between threads and as a staging area for data isn't meant to be processed yet.

**Comment by josephlewis42** [comment]
Hi @colinsurprenant 

I think it was a mistake in the original implementation. It makes sense if you assume you'll only produce a file every `@uploader_interval_seconds` but that's not always true. As an extreme example, on startup the plugin tries to recover old files, if we have ten of those and an upload_interval_sec of 1 hour, there'll be a perpetual 11 hour delay on data uploads.

I capped the other one at 60 seconds mostly because I don't trust the event ordering in the queue and it seemed like a reasonable compromise between resource usage and good user experience.

As a future enhancement I'd like to make the upload queue only hold things that are ready to be uploaded, that would help get rid of this weird logic.

Cheers!

**Comment by josephlewis42** [comment]
Any further thoughts/blockers for this? I'm willing to remove the min that sleep timer if it's causing concern.

Thanks!

**Comment by rosbo** [comment]
Woah. This logic is complex for no reason... The longer term solution is to rewrite it to use the streaming API and write data as we get it or only add the file to the `upload_queue` once it has been rolled-over.

I am ok with submitting this patch as a short term solution but we should definitely plan to implement the longer term solution.
